text/0000-read-by-prefix.md

```diff
@@ -0,0 +1,111 @@
+- Feature Name: read_by_prefix
+- Start Date: 2018-12-19
+- RFC PR:
+- Sawtooth Issue:
+
+# Summary
+[summary]: #summary
+
+This RFC proposes an additional API to transaction context,
+context.listAddresses(address_prefix), which will get address prefix as input
+and return list of all existing addresses that starts with the prefix.
+
+This capability exists when reading state from rest-api but is missing when 
+transaction processor reads state.
+
+# Motivation
+[motivation]: #motivation
+
+The motivation came when looking for a way to improve processing transaction 
+logic by making better use optimization that comes from using the radix tree. 
+
+Since addresses are stored in a radix tree, reading many addresses by 
+one prefix is much faster than reading the same amount of addresses one by one.
+Further more, read by prefix doesn't require transaction processor to know 
+which addresses exist, removing the need to store list of existing addresses 
+in a separate address.
+
+Having this capability allows a better and smarter address mapping this a 
+better transaction processor logic.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Get state by prefix is a feature available for reading via sawtooth rest-api.
+This RFC adds method to expose this capability of reading address prefix 
+to transaction processor.
+
+No change is recommended to the existing methods, so the implementation of this 
+RFC should not cause any breakage for pre-existing transaction processors.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Example of implementation can be found here: 
+Sawtooth-core + Python SDK changes: 
+https://github.com/arsulegai/sawtooth-core/tree/jira_1375 
+(commits: 4225dc7 and 2c3a223)
+
+Sawtooth-sdk-cxx changes: 
+https://github.com/arsulegai/sawtooth-sdk-cxx/tree/jira_1375 (commit: 23697e8)
+
+In the suggested implementation there are two proto changes:
+
+Addition to state_context.proto:
+
+
+	// A request from a handler/tp for list of addresses with valid data
+	message TpStateAddressesListRequest {
+		// The context id that references a context in the contextmanager
+		string context_id = 1;
+		repeated string addresses = 2;
```

Comment:
What if this request was for an address space with an exceptionally large number of addresses? Should they just be returned, or is there an upper limit, where the validator will send an error back with "request denied". If there should be an upper limit, how would we manage it?

Another approach may be a paging approach, where the request specifies the maximum number of addresses to return (with a max page size, perhaps configurable per-validator).

Reply:
Same as regular read, there is no limitation of data size in one address, it is up the the TP developer to make sure he doesn't write too much data to one address. 
Since only requesting address list without addresses data this won't get too big that easily.
No objection to adding paging and max, only that it is not done in regular read address, added it as to unresolved questions section.

Reply:
It is not the same type of problem, because read/write are relatively efficient. Returning a list of addresses from the state tree will not be efficient as it will require a large number of underlying database reads to accomplish.

Adding an on-chain setting for maximum size of address storage would probably be a good feature too.